"""
==============================================================================
ATTACK CASE: keh493_attackcase2.r2py
==============================================================================
Target NetID:        keh493
Original Test:       test21_filename_edge_cases.r2py
Test File Path:      submit/general_tests/test21_filename_edge_cases.r2py

RUNNER INFORMATION:
  Runner ID:         runner-3371
  Task UUID:         eaf428e2-0e54-4da9-b0dc-ace1deab602f
  Machine Type:      n2-standard-4
  Execution Server:  http://100.88.83.27:8000

EXECUTION INFORMATION:
  Status:            FAILED
  Start Time:        2025-10-27T19:23:27.174337
  End Time:          2025-10-27T19:23:32.377455
  Runtime:           5.2031 seconds
  Exit Code:         143

FILE HASHES:
  Monitor MD5:       977f7b2f65cb90c86022a50335e3c4df
  Attack MD5:        ff058ba3b8ff9df14934fbafc02bcb42
  Combined MD5:      f2d4fd41db6bc68092bcd89183455fd0

FAILURE DETAILS:
  Reason:            ---
Uncaught exception!
---
Following is a full traceback, and a user traceback.
The user traceback 

STDOUT:
  ---
  Uncaught exception!
  ---
  Following is a full traceback, and a user traceback.
  The user traceback excludes non-user modules. The most recent call is displayed last.
  Full debugging traceback:
    "/app/repy.py", line 134, in execute_namespace_until_completion
    "/app/virtual_namespace.py", line 116, in evaluate
    "/app/safe.py", line 610, in safe_run
    "/app/encasementlib.r2py", line 240, in <module>
    "/app/encasementlib.r2py", line 225, in secure_dispatch
    "/app/namespace.py", line 957, in __do_func_call
    "/app/namespace.py", line 1219, in wrapped_function
    "/app/virtual_namespace.py", line 116, in evaluate
    "/app/safe.py", line 610, in safe_run
    "monitor.r2py", line 126, in <module>
    "/app/encasementlib.r2py", line 175, in _secure_dispatch_closure
    "/app/encasementlib.r2py", line 225, in secure_dispatch
    "/app/namespace.py", line 957, in __do_func_call
    "/app/namespace.py", line 1219, in wrapped_function
    "/app/virtual_namespace.py", line 116, in evaluate
    "/app/safe.py", line 610, in safe_run
    "attack.r2py", line 29, in <module>
    "wrapper.r2py", line 255, in _wrapped_objc
    "wrapper.r2py", line 202, in _wrapped_func
    "monitor.r2py", line 76, in LPopenfile
    "monitor.r2py", line 14, in __init__
  User traceback:
    "/app/encasementlib.r2py", line 240, in <module>
    "/app/encasementlib.r2py", line 225, in secure_dispatch
    "monitor.r2py", line 126, in <module>
    "/app/encasementlib.r2py", line 175, in _secure_dispatch_closure
    "/app/encasementlib.r2py", line 225, in secure_dispatch
    "attack.r2py", line 29, in <module>
    "wrapper.r2py", line 255, in _wrapped_objc
    "wrapper.r2py", line 202, in _wrapped_func
    "monitor.r2py", line 76, in LPopenfile
    "monitor.r2py", line 14, in __init__
  Exception (with class 'exception_hierarchy.RepyArgumentError'): Cannot create explicit version files
  ---

ORIGINAL TEST DESCRIPTION:
  Targets: kl5058, npj2009
  Vulnerability: Edge cases in filename handling (length, parsing)
  Attack: Test with long filenames and files with multiple .v patterns

==============================================================================
"""

# ============================================================================
# REFERENCE MONITOR CODE (Target: keh493)
# ============================================================================
# TYPE = "type"
# ARGS = "args"
# RETURN = "return"
# EXCP = "exceptions"
# TARGET = "target"
# FUNC = "func"  
# OBJC = "objc"
# 
# 
# class VMFile():
#     def __init__(self, filename, create):
#         if create:
#             if '.v' in filename:
#                 raise RepyArgumentError("Cannot create explicit version files")
#             if filename in listfiles():
# 
#                 vNumber = 1
# 
#                 all_files = listfiles()
# 
#                 all_files.sort()
# 
#                 for i in all_files:
#                     if (filename + '.v') in i:
#                         vNumber += 1
# 
#                 new_name = filename + '.v' + str(vNumber)
# 
#                 if vNumber == 1:
#                     prev_file = openfile(filename, False)
#                 else:
#                     prev_file = openfile((filename +'.v' + str(vNumber-1)),False)
# 
#                 content = prev_file.readat(None,0)
# 
#                 self.VMfile = openfile(new_name, True)
#                 self.VMfile.writeat(content, 0)
#                 self.filename = new_name
#             else:
#                 self.VMfile = openfile(filename, True)
#                 self.filename = filename
#         else:
#             try:
#                 self.VMfile = openfile(filename, False)
#                 self.filename = filename
#             except FileNotFoundError:
#                 raise FileNotFoundError
# 
#     def readat(self, num_bytes, offset):
#         return self.VMfile.readat(num_bytes, offset)
# 
#     def writeat(self, data, offset): 
#         if '.v' not in self.filename:
#             if (self.filename + '.v1') in listfiles():
#                 raise FileInUseError
#             else:
#                 return self.VMfile.writeat(data, offset)
#         else:
#             index = 0
#             for i in self.filename:
#                 if i != '.':
#                     index +=1
#                 else:
#                     break
#             vNumber = int(self.filename[(index+2):])
#             if (self.filename[:index] + '.v' + str(vNumber + 1)) in listfiles():
#                 raise FileInUseError
#             else:
#                 return self.VMfile.writeat(data, offset)
# 
#     def close(self):
#         return self.VMfile.close()
# 
# 
# def LPopenfile(filename, create):
#     return VMFile(filename, create)
# 
# def LPremovefile(filename):
#     raise RepyArgumentError
# 
# def LPlistfiles():
#     actual_list = []
#     for i in listfiles():
#         if '.v' in i:
#             continue
#         else:
#             actual_list.append(i)
#     return actual_list
# 
# 
# # The code below sets up type checking and variable hiding for you.
# # You should not change anything below this point.
# sec_file_def = {
#     "obj-type": VMFile,
#     "name": "VMFile",
#     "writeat": {"type": "func", "args": (str, (int, long)), "exceptions": Exception, "return": (int, type(None)), "target": VMFile.writeat},
#     "readat": {"type": "func", "args": ((int, long, type(None)), (int, long)), "exceptions": Exception, "return": str, "target": VMFile.readat},
#     "close": {"type": "func", "args": None, "exceptions": Exception, "return": (bool, type(None)), "target": VMFile.close}
# }
# 
# CHILD_CONTEXT_DEF["openfile"] = {
#     TYPE: OBJC,
#     ARGS: (str, bool),
#     EXCP: Exception,
#     RETURN: sec_file_def,
#     TARGET: LPopenfile
# }
# 
# CHILD_CONTEXT_DEF["removefile"] = {
#     TYPE: FUNC,
#     ARGS: (str,),
#     EXCP: Exception,
#     RETURN: type(None),
#     TARGET: LPremovefile
# }
# 
# CHILD_CONTEXT_DEF["listfiles"] = {
#     TYPE: FUNC,
#     ARGS: None,
#     EXCP: Exception,
#     RETURN: list,
#     TARGET: LPlistfiles
# }
# 
# # Execute the user code
# secure_dispatch_module()
# ============================================================================

# Test 1: Long filename
longname = "a" * 100
f = openfile(longname, True)
f.writeat("data", 0)
f.close()

# Create version
f = openfile(longname, True)
content = f.readat(None, 0)
if content != "data":
    log("ERROR: Version should copy content\n")
    exitall()

# Should be writable since it's the latest version
try:
    f.writeat("new", 0)
except FileInUseError:
    log("ERROR: Latest version should be writable\n")
    exitall()
f.close()

# Test 2: File with .v but not a version
f2 = openfile("data.value", True)
f2.writeat("value_data", 0)
f2.close()

# Verify it appears in listfiles
files = listfiles()
if "data.value" not in files:
    log("ERROR: data.value should be in listfiles\n")
    exitall()
